Fixes #403

dekobon opened the pull request:
The exception type returned when downloading a dir as a stream is too generic.

We now throw a specific exception type when we encounter states where the
wrong object type is encountered.

dekobon commented on java-manta-client-unshaded/src/main/java/com/joyent/manta/exception/MantaUnexpectedObjectTypeException.java:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @param responseHeaders headers object to attach to exception
+     */
+    public void setResponseHeaders(final MantaHttpHeaders responseHeaders) {
```
This method is really nice because it allows consumers to salvage information about the request out of the failure mode.

tjcelaya commented:
I'd like to continue shifting our testing efforts to unit tests. Please add the following file at `java-manta-client-unshaded/src/test/java/com/joyent/manta/client/MantaDirectoryListingIteratorTest.java`:

```java
package com.joyent.manta.client;

import com.joyent.manta.exception.MantaUnexpectedObjectTypeException;
import com.joyent.manta.http.HttpHelper;
import com.joyent.manta.http.MantaHttpRequestFactory;
import org.apache.http.Header;
import org.apache.http.HttpEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.entity.ContentType;
import org.apache.http.message.BasicHeader;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import org.testng.Assert;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import static com.joyent.manta.client.MantaDirectoryListingIterator.MAX_RESULTS;
import static org.apache.http.HttpHeaders.CONTENT_TYPE;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

public class MantaDirectoryListingIteratorTest {

    @Mock
    private HttpHelper httpHelper;

    @Mock
    private CloseableHttpResponse response;

    @Mock
    private HttpEntity responseEntity;

    @BeforeMethod
    public void setup() throws Exception{
        MockitoAnnotations.initMocks(this);

        when(httpHelper.getRequestFactory()).thenReturn(new MantaHttpRequestFactory("http://localhost"));
        when(httpHelper.executeRequest(any(), any())).thenReturn(response);
        when(response.getEntity()).thenReturn(responseEntity);
    }

    @AfterMethod
    public void teardown() throws Exception {
        Mockito.validateMockitoUsage();
    }

    // uncomment once it's okay to call next before hasNext
    // @Test
    // public void allowsSkippingFirstResult() throws Exception {
    //     final MantaDirectoryListingIterator itr = new MantaDirectoryListingIterator("/usr/stor/dir", httpHelper, MAX_RESULTS);
    //
    //     Assert.assertNull(itr.next());
    // }

    @Test
    public void throwsAppropriateExceptionWhenListingObject() throws Exception {
        final String dirPath = "/user/stor/directory";
        final Header contentTypeHeader = new BasicHeader(CONTENT_TYPE, ContentType.APPLICATION_OCTET_STREAM.toString());

        when(responseEntity.getContentType()).thenReturn(contentTypeHeader);
        when(response.getFirstHeader(CONTENT_TYPE)).thenReturn(contentTypeHeader);

        // uncomment once it's okay to call next before hasNext
        // Assert.expectThrows(MantaUnexpectedObjectTypeException.class, () ->
        //         new MantaDirectoryListingIterator(dirPath, httpHelper, MAX_RESULTS).next());

        Assert.expectThrows(MantaUnexpectedObjectTypeException.class, () ->
                new MantaDirectoryListingIterator(dirPath, httpHelper, MAX_RESULTS).hasNext());
    }

}
```

dekobon commented:
I've updated the PR to incorporate the following:

* The patchset from @tjcelaya 
* A fix for `MantaDirectoryListingIterator` being dependent on calling `hasNext()` before `next()`
* Converted new integration tests to unit tests.

Please review the new changes.
